File: others/string-concat/cpp/bench_strcat.cpp
/**
 * $Release: $
 * $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
 * $License: Creative Commons Attribution (CC BY) $
 */

/*
 *  Are you a C++ expert? PLEASE ADVICE ME!
 */


#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <vector>
#include <iostream>
//#include <time.h>
#include <sys/time.h>


const char *members[] = {"Haruhi", "Mikuru", "Yuki", "Itsuki", "Kyon"};


void task_empty(int loop) {
    const char *s1 = members[0], *s2 = members[1], *s3 = members[2], *s4 = members[3], *s5 = members[4];
    while (--loop >= 0) {
        std::string buf("");
        for (int j = 0; j < 20; j++) {
            // pass
        }
    }
}


void task_strcat1(int loop) {
    const char *s1 = members[0], *s2 = members[1], *s3 = members[2], *s4 = members[3], *s5 = members[4];
    while (--loop >= 0) {
        std::string buf("");
        //
        buf.append("<table>\n");
        for (int j = 0; j < 20; j++) {
            buf.append("  <tr>\n"
"    <td>"); buf.append(s1); buf.append("</td>\n"
"    <td>"); buf.append(s2); buf.append("</td>\n"
"    <td>"); buf.append(s3); buf.append("</td>\n"
"    <td>"); buf.append(s4); buf.append("</td>\n"
"    <td>"); buf.append(s5); buf.append("</td>\n"
"  </tr>\n");
        }
        buf.append("</table>\n");
        //
        //if (loop == 1) std::cout << buf;
    }
}


void task_strcat2(int loop) {
    const char *s1 = members[0], *s2 = members[1], *s3 = members[2], *s4 = members[3], *s5 = members[4];
    while (--loop >= 0) {
        std::string buf("");
        buf.reserve(4*1024);       // !!!
        //
        buf.append("<table>\n");
        for (int j = 0; j < 20; j++) {
            buf.append("  <tr>\n"
"    <td>"); buf.append(s1); buf.append("</td>\n"
"    <td>"); buf.append(s2); buf.append("</td>\n"
"    <td>"); buf.append(s3); buf.append("</td>\n"
"    <td>"); buf.append(s4); buf.append("</td>\n"
"    <td>"); buf.append(s5); buf.append("</td>\n"
"  </tr>\n");
        }
        buf.append("</table>\n");
        //
        //if (loop == 1) std::cout << buf;
    }
}


void task_vector1(int loop) {
    const char *s1 = members[0], *s2 = members[1], *s3 = members[2], *s4 = members[3], *s5 = members[4];
    while (--loop >= 0) {
        std::vector<const char*> buf;
        //
        buf.push_back("<table>\n");
        for (int j = 0; j < 20; j++) {
            buf.push_back("  <tr>\n"
"    <td>"); buf.push_back(s1); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s2); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s3); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s4); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s5); buf.push_back("</td>\n"
"  </tr>\n");
        }
        buf.push_back("</table>\n");
        //
        //if (loop == 1) std::cout << buf;
        //printf("*** debug: buf.size()=%d\n", buf.size());
    }
}


void task_vector2(int loop) {
    const char *s1 = members[0], *s2 = members[1], *s3 = members[2], *s4 = members[3], *s5 = members[4];
    while (--loop >= 0) {
        std::vector<const char*> buf(256);
        //
        buf.push_back("<table>\n");
        for (int j = 0; j < 20; j++) {
            buf.push_back("  <tr>\n"
"    <td>"); buf.push_back(s1); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s2); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s3); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s4); buf.push_back("</td>\n"
"    <td>"); buf.push_back(s5); buf.push_back("</td>\n"
"  </tr>\n");
        }
        buf.push_back("</table>\n");
        //
        //if (loop == 1) std::cout << buf;
        //printf("*** debug: buf.size()=%d\n", buf.size());
    }
}


double run(const char *title, void (*body)(int loop_), int loop, double empty_time) {
    struct timeval start, stop;
    gettimeofday(&start, NULL);
    (*body)(loop);
    gettimeofday(&stop, NULL);
    int sec  = stop.tv_sec  - start.tv_sec;
    int usec = stop.tv_usec - start.tv_usec;
    double real = (double)sec + (double)usec / 1000000.0;
    double actual = real - empty_time;
    printf("%-35s %10.3f %10.3f\n", title, real, actual);
    return real;
}


int main(int argc, char *argv[]) {
    int loop = 100*1000;
    int cycle = 3;
    char *s;
    if ((s = getenv("N")) != NULL) loop  = atoi(s);
    if ((s = getenv("C")) != NULL) cycle = atoi(s);
    //
    printf("# loop = %d\n", loop);
    double empty_time = 0.0;
    for (int i = 1; i <= cycle; i++) {
        printf("\n");
        printf("# cycle=%d %25s %10s %10s\n", i, " ", "real", "actual");
        empty_time = \
        run("(Empty)",                 task_empty,   loop, empty_time);
        run("string().append()",       task_strcat1, loop, empty_time);
        run("string(4*1024).append()", task_strcat2, loop, empty_time);
        run("vector.push_back() (no join)",       task_vector1, loop, empty_time);
        run("vector(256).push_back() (no join)",  task_vector2, loop, empty_time);
    }
    //
    return 0;
}

File: others/string-concat/python/bench_strcat.py
# -*- coding: utf-8 -*-

###
### $Release: $
### $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
### $License: Creative Commons Attribution (CC BY) $
###

from __future__ import with_statement

import sys, os, time
try:
    from cStringIO import StringIO
except ImportError:
    from io import StringIO

try:
    xrange
except NameError:
    xrange = range

from benchmarker import Benchmarker

members = ["Haruhi", "Mikuru", "Yuki", "Itsuki", "Kyon",]
s1, s2, s3, s4, s5 = members


with open('../expected.out') as f:
    expected = f.read()

def verify(out):
    assert expected == out


loop = int(os.getenv('N') or 100*1000)
if os.getenv('C'):
    cycle = int(os.getenv('C'))
else:
    cycle = 1


for bm in Benchmarker(cycle=cycle):
    print("## loop = %s" % loop)

    with bm.empty():
        for i in xrange(loop):
            for j in xrange(20):
                pass


    with bm("list.append()"):
        for i in xrange(loop):
            buf = []
            buf.append('''<table>\n''')
            for j in xrange(20):
                buf.append('''  <tr>
    <td>'''); buf.append(s1); buf.append('''</td>
    <td>'''); buf.append(s2); buf.append('''</td>
    <td>'''); buf.append(s3); buf.append('''</td>
    <td>'''); buf.append(s4); buf.append('''</td>
    <td>'''); buf.append(s5); buf.append('''</td>
  </tr>\n''');
            buf.append('''</table>\n''')
            out = "".join(buf)
    verify(out)


    with bm("append()"):
        for i in xrange(loop):
            buf = []; append = buf.append
            append('''<table>\n''')
            for j in xrange(20):
                append('''  <tr>
    <td>'''); append(s1); append('''</td>
    <td>'''); append(s2); append('''</td>
    <td>'''); append(s3); append('''</td>
    <td>'''); append(s4); append('''</td>
    <td>'''); append(s5); append('''</td>
  </tr>\n''');
            append('''</table>\n''')
            out = "".join(buf)
    verify(out)


    with bm("list.extend()"):
        for i in xrange(loop):
            buf = []
            buf.extend(('''<table>\n''',))
            for j in xrange(20):
                buf.extend(('''  <tr>
    <td>''', s1, '''</td>
    <td>''', s2, '''</td>
    <td>''', s3, '''</td>
    <td>''', s4, '''</td>
    <td>''', s5, '''</td>
  </tr>\n''',))
            buf.extend(('''</table>\n''',))
            out = "".join(buf)
    verify(out)


    with bm("extend()"):
        for i in xrange(loop):
            buf = []; extend = buf.extend
            extend(('''<table>\n''',))
            for j in xrange(20):
                extend(('''  <tr>
    <td>''', s1, '''</td>
    <td>''', s2, '''</td>
    <td>''', s3, '''</td>
    <td>''', s4, '''</td>
    <td>''', s5, '''</td>
  </tr>\n''',))
            extend(('''</table>\n''',))
            out = "".join(buf)
    verify(out)


    with bm("extend() only (no join)"):
        for i in xrange(loop):
            buf = []; extend = buf.extend
            extend(('''<table>\n''',))
            for j in xrange(20):
                extend(('''  <tr>
    <td>''', s1, '''</td>
    <td>''', s2, '''</td>
    <td>''', s3, '''</td>
    <td>''', s4, '''</td>
    <td>''', s5, '''</td>
  </tr>\n''',))
            extend(('''</table>\n''',))
            #out = "".join(buf)
    #verify(out)


    with bm("extend()+str()"):
        for i in xrange(loop):
            buf = []; extend = buf.extend
            extend(('''<table>\n''',))
            for j in xrange(20):
                extend(('''  <tr>
    <td>''', str(s1), '''</td>
    <td>''', str(s2), '''</td>
    <td>''', str(s3), '''</td>
    <td>''', str(s4), '''</td>
    <td>''', str(s5), '''</td>
  </tr>\n''',))
            extend(('''</table>\n''',))
            out = "".join(buf)
    verify(out)


    with bm("_str=str;extend()+_str()"):
        _str = str
        for i in xrange(loop):
            buf = []; extend = buf.extend
            extend(('''<table>\n''',))
            for j in xrange(20):
                extend(('''  <tr>
    <td>''', _str(s1), '''</td>
    <td>''', _str(s2), '''</td>
    <td>''', _str(s3), '''</td>
    <td>''', _str(s4), '''</td>
    <td>''', _str(s5), '''</td>
  </tr>\n''',))
            extend(('''</table>\n''',))
            out = "".join(buf)
    verify(out)


    with bm("buf[-1:] = ()"):
        for i in xrange(loop):
            buf = [""]
            buf[-1:] = ('''<table>\n''', "")
            for j in xrange(20):
                buf[-1:] = ('''  <tr>
    <td>''', s1, '''</td>
    <td>''', s2, '''</td>
    <td>''', s3, '''</td>
    <td>''', s4, '''</td>
    <td>''', s5, '''</td>
  </tr>\n''', "")
            buf[-1:] = ('''</table>\n''', "")
            out = "".join(buf)
    verify(out)


    with bm("StringIO.write()"):
        for i in xrange(loop):
            buf = StringIO()
            write = buf.write
            write('''<table>\n''')
            for j in xrange(20):
                write('''  <tr>
    <td>'''); write(s1); write('''</td>
    <td>'''); write(s2); write('''</td>
    <td>'''); write(s3); write('''</td>
    <td>'''); write(s4); write('''</td>
    <td>'''); write(s5); write('''</td>
  </tr>\n''');
            write('''</table>\n''')
            out = buf.getvalue()
    verify(out)
